woocommerce/woocommerce
# Bump WooCommerce minimum required PHP version to 7.2

### All Submissions:


### Note for reviewers

It's recommended to review each commit individually as they are pretty much orthogonal. Also although there are 100+ changed files, most of the changes are in two of the unit test related commits and are quite repetitive code additions or replacements.

### Changes proposed in this Pull Request

[As announced back in January](https://developer.woocommerce.com/2022/01/05/new-requirement-for-woocommerce-6-5-php-7-2/) the minimum PHP version required will be 7.2 as of WooCommerce 6.5. This pull request makes effective this transition and adjusts dependencies and unit tests as needed. Note that there's also a breaking change related to the Emogrifier package and the `woocommerce_emogrifier` hook.

#### Core changes and package updates

* _Requires PHP_ is changed in `woocommerce.php`
* _require-php_ and _config-platform-php_ is changed in `composer.json`
* Dependency versions update in `composer.json`:
  * The _pelago/emogrifier_ package version is changed to `^6.0`
  * The _maxmind-db/reader_ package version is bumped to `^1.6`
  * The _composer/installers_ package version is bumped to `^1.9` (a bump to 2.0 not possible due to conflict with Admin package)
  * The required version for _phpunit_ is now explicitly indicated as 7.5.20, otherwise it will install v8 as it's the newerst version that supports PHP 7.2.

#### Changes in unit tests

Some of the changes are for making the unit tests ready for PHPUnit 8, which is what WooCommerce will move to once the minimum supported WordPress version is 5.9 ([details here](https://make.wordpress.org/core/handbook/references/phpunit-compatibility-and-wordpress-versions/)).

* PHP 7.0 and 7.1 are removed from the list of PHP versions to run the unit tests in CI.
* Add explicit `void` return types to `setUp` and `tearDown` methods in unit test classes.
* Fix calls to deprecated PHPUnit methods, e.g. `assertInternalType` are replaced with `assertIs(type)`.
* Add [even more](https://github.com/woocommerce/woocommerce/pull/31286) `#[\ReturnTypeWillChange]` attributes, these are required for PHP 8.1.

#### Breaking change

WooCommerce has a `woocommerce_emogrifier` action that can be used to tweak how the emails WooCommerce sends are styled. Until now this action got an object of type `Emogrifier` as a parameter, the problem is that as of v4 of the Emogrifier package this class no longer exists, and in order to be compatible with PHP 8 we need at least v5 of the package.

The workaround in this pull request is to pass an instance of the `CssInliner` class, which is the replacement for the `Emogrifier` class. The new class keeps some of the methods of the old class but not all of them, hence the breaking change.

For reference this is the list of the `Emogrifier` methods that are still present in `CssInliner`:

* addAllowedMediaType
* addExcludedSelector
* disableInlineStyleAttributesParsing
* removeAllowedMediaType
* removeExcludedSelector
* setDebug

#### Other changes

A visibility modifier (`public` or `private`) has been added to the handful of constants defined in the classes in `src`, visibility modifier for constants is a feature introduced in PHP 7.2 that improves the code readability. The code in `src/Admin` and `src/Internal/Admin`  has been excluded frrom this refactor since there's a lot of files here and that deserves a separate project.

### How to test the changes in this Pull Request:

Test that WooCommerce works and the unit tests still pass (TODO: this could probably be improved)

One of the issues that triggered the bump to PHP 7.2 was [the impossibility to install WooCommerce as a Composer package in PHP 8](https://github.com/woocommerce/woocommerce/issues/30027). To test that this has been fixed:

1. Create an empty directory with any name.
2. Inside that directory create a `composer.json` file with this content (adjust "url" to the location of your local copy of WooCommerce, but remember that it must end with `/plugins/woocommerce`):

```json
{
    "name": "foobar/fizzbuzz",
    "minimum-stability": "dev",
    "repositories": [
      {
          "type": "path",
          "url": "../woocommerce/plugins/woocommerce"
      }
    ],
    "require": {
        "woocommerce/woocommerce": "*"
    }
}
```

3. Verify that `composer install` succeeds for all versions of PHP from 7.2 to 8.1 when your local copy of WooCommerce is in the branch of this pull request (in `trunk` it will fail in PHP 8.x).

### Other information:




### Changelog entry

> Tweak - Bump minimum required PHP version of WooCommerce to 7.2

### FOR PR REVIEWER ONLY:
